src/cockpit/samples.py

```diff
@@ -192,49 +192,55 @@ class CGroupSampler(Sampler):
     cgroups_v2: Optional[bool] = None
     cgroups_v2_path = '/sys/fs/cgroup/'
 
-    def read_cgroup_keyed_stat(self, samples, path, cgroup, name, key):
-        with open(path) as stat:
-            for line in stat:
-                if not line.startswith(key):
-                    continue
-
-                value = int(line.split()[-1])
-                if sys.maxsize > value > 0:
-                    samples[name][cgroup] = value / 1000
-
-    def read_cgroup_integer_stat(self, samples, path, cgroup, name):
+    @staticmethod
+    def read_cgroup_integer_stat(rootfd, statfile, include_zero=False, key=b''):
         # Not every stat is available, such as cpu.weight
         try:
-            with open(path) as stat:
-                # Some samples such as "memory.max" contains "max" when there is a no limit
-                try:
-                    value = int(stat.read().strip())
-                except ValueError:
-                    return
-
-                if sys.maxsize > value > 0:
-                    samples[name][cgroup] = value
-        except FileNotFoundError:
-            pass
+            fd = os.open(statfile, os.O_RDONLY, dir_fd=rootfd)
+        except Exception:
+            return
+
+        data = os.read(fd, 1024)
```

Comment:
If this fails for whatever reason, you'll leak `fd`.  Please put the `close()` in a `finally:` block.

Reply:
Sure, but that doesn't fix the potential exception right? Although that seems pretty unlikely.